Contract upload

jhughes982 opened the pull request:
## Works towards #6 #7 #8 #258

- This PR adds the ability to upload and download contracts,
- A frontend form has been added to allow pdf upload as well as contract metadata,
- Backend methods have been added for validating the contract metadata and integrating with S3 for pdf storage,
- I suspect this PR will generate more discussion around how all this works and what the final metadata fields are, thoughts are welcome,
- Tests will be added when this functionality is signed off

<!--
For example "Resolves #42" if this PR resolves issue 42
-->

### Checklist

- [ ] Documentation has been updated

t-young31 commented on internal/handler/web/studies.go:
```diff
@@ -70,6 +72,19 @@ func assetToOpenApiAsset(asset types.Asset) openapi.Asset {
 
+func contractToOpenApiContract(contract types.Contract) openapi.Contract {
+	return openapi.Contract{
+		Id:                    contract.ContractID.String(),
```
hmm. I was expecting the database Id and the object id to be the same. why should they be different?

jhughes982 commented on internal/types/studies.go:
```diff
@@ -96,3 +98,31 @@ type AssetLocation struct {
+	User  User  `gorm:"foreignKey:UploadedBy"`
+}
+
+// used for parsing multipart form data when uploading a contract (before it is stored)
+type ContractFormData struct {
+	OrganisationSignatory string                `form:"organisation_signatory"`
+	ThirdPartyName        string                `form:"third_party_name"`
+	Status                string                `form:"status"`
+	ExpiryDate            string                `form:"expiry_date"`
+	File                  *multipart.FileHeader `form:"file"`
```
Yeah I don't like it either. I'll have a play around to see what I can get working.

t-young31 replied:
fair enough. I was thinking you'd just not do the rollback and allow the object to not exist (throwing an error of course), then if it does the user can retry the upload. whether that's sensible or not maybe depends if we're going to allow the contract to be mutable (i.e. can upload a new object with the same metadata)

t-young31 replied:
nope – let's come back to it in future if needed

jhughes982 commented on internal/service/studies/contracts.go:
```diff
@@ -2,24 +2,86 @@ package studies
 )
 
+func (s *Service) ValidateContractMetadata(organisationSignatory string, thirdPartyName string, status string, expiryDateStr string) *openapi.ContractUploadValidationError {
+	if !validation.ContractNamePattern.MatchString(strings.TrimSpace(organisationSignatory)) {
```
[remove trimspace](https://github.com/ucl-arc-tre/portal/pull/295/commits/5b2d69c8a4684ef7af68dbbd371f438ef444d8a5)

jhughes982 commented:
Just a heads up. I've experienced some variable behaviour with the S3 upload during local testing. A few times now I've had either the upload or the download not work (e.g. it says the bucket doesn't exist) only for it to work again after i tear down the containers and spin them up again.

I can't rule out that this was caused by me somehow, so let me know if anyone experiences this too.

t-young31 commented on internal/service/studies/contracts.go:
```diff
@@ -2,24 +2,87 @@ package studies
+		}
+	}
+
+	if status != "proposed" && status != "active" && status != "expired" {
```
better to match enum probably

jhughes982 replied:
[now using typed enum](https://github.com/ucl-arc-tre/portal/pull/295/commits/26e98fc00f5f21eab8a7082d9e0d355308af5310)

jhughes982 commented on internal/types/studies.go:
```diff
@@ -96,3 +96,21 @@ type AssetLocation struct {
+// Contract represents a PDF contract document associated with an asset
+type Contract struct {
+	ModelAuditable
+	StudyID               uuid.UUID `gorm:"not null;index"`
```
[removed studyId from contract type](https://github.com/ucl-arc-tre/portal/pull/295/commits/1eb584775ea70cf59af41a0cdfd809e38cb5a246)

jhughes982 commented on internal/handler/web/studies.go:
```diff
@@ -273,21 +288,87 @@ func (h *Handler) GetStudiesStudyIdAgreements(ctx *gin.Context, studyId string)
+		return
+	}
+
+	expiryDate, err := time.Parse("2006-01-02", formData.ExpiryDate)
```
[moving date validation to config](https://github.com/ucl-arc-tre/portal/pull/295/commits/75bb638b2efe7b6ed6133f85d6941f8d9ab5e8ff)

jhughes982 commented:
I've addressed all the comments now (just waiting for replies on a few of them). I think I'll stop here for now and have a look at #294 which we agreed should go in before this PR.

acholyn commented on web/src/components/assets/ContractManagement.tsx:
```diff
@@ -1,134 +1,93 @@
+      if (response.response.ok && response.data) {
+        setContracts(response.data);
+      } else {
+        throw new Error(`Failed to fetch contracts: ${response.response.status} ${response.response.statusText}`);
```
this keeps getting thrown for me after creating an asset 🤔

acholyn commented on web/src/components/assets/ContractUploadForm.tsx:
```diff
@@ -0,0 +1,268 @@
+                maxLength: { value: 100, message: "Organisation Signatory must be less than 100 characters" },
+              })}
+              className={styles["form-input"]}
+              placeholder="Enter organisation signatory name"
```
might be worth adding an eg. I'm not sure if this is meant to be a person's name at the org or the org itself

t-young31 replied:
> but is it in principle possible to construct queries to retrieve all contracts and assets associated with Studies that a particular IAO owns?

yeah, no sweat

jhughes982 replied:
> but is it in principle possible to construct queries to retrieve all contracts and assets associated with Studies that a particular IAO owns

Yes we can do this. Do you need this to be built into this PR (will need UI/UX considerations). Or is this something just to keep in mind for future functionality?

jhughes982 commented on internal/types/studies.go:
```diff
@@ -96,3 +96,19 @@ type AssetLocation struct {
+
+	// Relationships
+	Asset Asset `gorm:"foreignKey:AssetID"`
+	User  User  `gorm:"foreignKey:CreatorUserID"`
```
[updated type name](https://github.com/ucl-arc-tre/portal/pull/295/commits/e6ef0ed082ad6ff429bf7f816de2ed72e3513b63)

jhughes982 commented on internal/service/studies/contracts.go:
```diff
@@ -2,29 +2,95 @@ package studies
+			ErrorMessage: "Invalid expiry date format",
+		}
+	}
+
```
[now validating pdf type](https://github.com/ucl-arc-tre/portal/pull/295/commits/bb20694cf3683f2e0cb5d2ef654cbf166f11f2c7)

jhughes982 commented on internal/handler/web/studies.go:
```diff
@@ -273,21 +288,98 @@ func (h *Handler) GetStudiesStudyIdAgreements(ctx *gin.Context, studyId string)
+		ExpiryDate:            expiryDate,
+	}
+
+	pdfContractObj := types.S3Object{
```
[updating variable name](https://github.com/ucl-arc-tre/portal/pull/295/commits/5530ddb6ae9d29298075bba3a95ffdfaf3b53d99)

jhughes982 commented on api.web.yaml:
```diff
@@ -955,11 +978,70 @@ components:
+          description: Current status of the contract
+        expiry_date:
+          type: string
+          description: Contract expiry date
```
[describing contract expiry format](https://github.com/ucl-arc-tre/portal/pull/295/commits/4ac00fdbb4a1835ce46b54873a2d286ec567af81)
